source/rust_verify/src/verifier.rs

```diff
@@ -591,16 +683,48 @@ impl Verifier {
                 let error_as = if recommends_rerun { ErrorAs::Note } else { ErrorAs::Error };
                 let s =
                     if recommends_rerun { "Function-Check-Recommends " } else { "Function-Def " };
-                let invalidity = self.run_commands_queries(
-                    compiler,
-                    error_as,
-                    air_context,
-                    &commands,
-                    &HashMap::new(),
-                    &snap_map,
-                    &(s.to_string() + &fun_as_rust_dbg(&function.x.name)),
-                );
-                if invalidity && !recommends_rerun && !self.args.no_auto_recommends_check {
+
+                let mut function_invalidity = false;
+                for command in commands.iter().map(|x| &*x) {
+                    let CommandsWithContextX { span, desc: _, commands: _, spinoff_z3 } =
+                        &**command;
+                    let command_invalidity: bool;
+                    if *spinoff_z3 {
+                        let mut spinoff_air_context = self.new_air_context_with_module_context(
+                            ctx,
+                            module,
+                            &(function.x.name).path,
+                            datatype_commands.to_vec(),
+                            function_decl_commands.to_vec(),
+                            function_spec_commands.to_vec(),
+                            function_axiom_commands.to_vec(),
+                            false,
+                            &span,
+                        )?;
+                        command_invalidity = self.run_commands_queries(
+                            compiler,
+                            error_as,
+                            &mut spinoff_air_context,
+                            command.clone(),
+                            &HashMap::new(),
+                            &snap_map,
+                            &(s.to_string() + &fun_as_rust_dbg(&function.x.name)),
+                        );
+                    } else {
+                        command_invalidity = self.run_commands_queries(
+                            compiler,
+                            error_as,
+                            air_context,
+                            command.clone(),
+                            &HashMap::new(),
+                            &snap_map,
+                            &(s.to_string() + &fun_as_rust_dbg(&function.x.name)),
+                        );
+                    }
```

Comment:
What about something like:
```rust
let mut spinoff_z3_context;
let query_air_context = if *spinoff_z3 {
    spinoff_z3_context = self.new_air_context_with_module_context(
        ctx,
        module,
        &(function.x.name).path,
        &*datatype_commands,
        &function_decl_commands,
        &function_spec_commands,
        &function_axiom_commands,
        recommends_rerun,
        &span,
    )?;
    &mut spinoff_z3_context
} else {
    &mut *air_context
};
command_invalidity = self.run_commands_queries(
    compiler,
    error_as,
    query_air_context,
    command.clone(),
    &HashMap::new(),
    &snap_map,
    &(s.to_string() + &fun_as_rust_dbg(&function.x.name)),
);
```
to avoid the redundant call?